### ISSUE TITLE
Class iterator with recursive generator throws Maximum call stack size exceeded

### ISSUE THREAD

**Issue by pgbross** [babel/babel#1251]
I have a class, Screengroup, that has a list containing Screens and/or Screengroups which implements the following ES6 style iterator that I use to flatten a hierarchy of objects:
```
[Symbol.iterator]() {
    let iterator = function*(tree) {
      if (Array.isArray(tree)) {
        for (let i = 0; i < tree.length; i++) {
          let item = tree[i]
          if (item instanceof Screen) {
            yield item
          } else {
            yield * iterator([...item.screenlist])
          }
        }
      } else if (tree instanceof Screengroup) {
        iterator([...tree.screenlist])
      }
    }
    return iterator([...this.screenlist])
  }
```
In the latest release of Babel (v5.1.2) I get an error ***'Uncaught RangeError: Maximum call stack size exceeded'*** in the runtime.js (the regenerator runtime).

if I change the implementation subtly to declare a variable separately from assigning a function to it, it behaves correctly, as in:

```
  [Symbol.iterator]() {
    let iterator  // <===== separate variable definition works here
    iterator = function*(tree) {
      if (Array.isArray(tree)) {
        for (let i = 0; i < tree.length; i++) {
          let item = tree[i]
          if (item instanceof Screen) {
            yield item
          } else {
            yield * iterator([...item.screenlist])
          }
        }
      } else if (tree instanceof Screengroup) {
        iterator([...tree.screenlist])
      }
    }
    return iterator([...this.screenlist])
  }
```

The generated javascript in both cases is quite complex to follow, but it appears it might be reusing the symbol iterator incorrectly in the failure example:

```
 {
    key: Symbol.iterator,
    value: function () {
      var iterator = (function (_iterator) {
        var iterator = regeneratorRuntime.mark(function iterator(_x) {  //<===== this reuse of the name iterator looks suspicious here
          return regeneratorRuntime.wrap(function iterator$(context$4$0) {
            while (1) switch (context$4$0.prev = context$4$0.next) {
   ... rest of the code not shown
``` 

I believe the first version shown above with the function assignment directly to the local variable used to work around about Babel v4.7, but I have not been able to revert to an older version and make it work again since upgrading to the latest Babel.

--philip

**Comment by sebmck** [comment]
Fixed as of 5.1.3, thanks!

**Comment by pgbross** [comment]
I have tested my code with Babel v5.1.8 and I get the same behaviour as originally described,  'Uncaught RangeError: Maximum call stack size exceeded' if I declare my iterator and assign the variable in the same line.

Comparing the generated code in the two cases a bit more carefully, the *failure* case includes this extra function (the first few lines are shown above):
```
value: function () {
      var iterator = (function (_iterator) { <==== this parameter is only referenced in the .toString() method below
        var marked3$0 = [iterator].map(regeneratorRuntime.mark); //<==== reusing the name iterator looks suspicious here

        function iterator(_x) {
          var args$4$0 = arguments;
          return regeneratorRuntime.wrap(function iterator$(context$4$0) {
            while (1) switch (context$4$0.prev = context$4$0.next) {
              case 0:
                return context$4$0.delegateYield(iterator.apply(this, args$4$0), 't0', 1);

              case 1:
                return context$4$0.abrupt('return', context$4$0.t0);

              case 2:
              case 'end':
                return context$4$0.stop();
            }
          }, marked3$0[0], this);
        }

        iterator.toString = function () {
          return _iterator.toString();
        };

        return iterator;
      })(regeneratorRuntime.mark(function callee$2$0(tree) {
        var i, item;  //<===== from here the generated code is almost identical
// rest of the generated code not shown
```

I note that issue [#1164](https://github.com/babel/babel/issues/1164) talks about a similar looking error with property-method assignements and wonder if there may be a related fault here?

--philip

**Comment by sebmck** [comment]
I thought it was because of the tail call recursion transformer but it's actually because of the function name inference.

**Comment by sebmck** [comment]
Can you provide a code example that I can actually execute? Hard to debug without being able to execute the code for this case. Thanks!

**Comment by pgbross** [comment]
I will try and get an example that demonstrates the issue that doesn't require large sections of the rest of the application.

**Comment by pgbross** [comment]
Here is a single file with the broken iterator generator, that contains a couple of simple classes with sufficient properties to let it work.
```
import '../public/js/runtime'
import '../public/js/shim.min'

class Screen {
  constructor(name) {
    this.name = name
  }
}

class Screengroup {
  constructor(name, list) {
    this.name = name
    this.screenlist = list
  }

  [Symbol.iterator]() {
    let // iterator   // <=== uncomment this to let it work
      iterator = function*(tree) {
        if (Array.isArray(tree)) {
          for (let i = 0; i < tree.length; i++) {
            let item = tree[i]
            if (item instanceof Screen) {
              yield item
            } else {
              yield * iterator([...item.screenlist])
            }
          }
        } else if (tree instanceof Screengroup) {
          iterator([...tree.screenlist])
        }
      }
    return iterator([...this.screenlist])
  }
}

class main {
  static run() {
    let list = [new Screen('one'), new Screengroup('group one', [])]
    let root = new Screengroup('root', list)
    let flat = [...root]

    console.dir(flat)
  }
}
main.run()
```

I saved this example as a single .js file, transpiled using babel v5.1.8 (--modules 'common'), and then ran it at a terminal prompt.

```
babel --modules 'common' fault.js > f.js
iojs f
```

Hope this helps tracking down the issue,

--philip

**Comment by sebmck** [comment]
Fixed. I'll push out a patch release.

**Comment by sebmck** [comment]
Fixed as of 5.1.9, thanks!
